File: src/com/lifx/Messages/Device/StateLocation.java
package com.lifx.Messages.Device;

import java.math.BigInteger;
import java.time.Instant;
import java.util.Random;

import com.lifx.Messages.DataTypes.Payload;

public class StateLocation extends Payload {
	int code = 50;
	byte[] location;			// 16-Bytes
	String label;				// 32-Bytes
	BigInteger updated_at;		// 64-Bits (Unsigned)
	
	public StateLocation() {
		location = new byte[16];
		new Random().nextBytes(location);
		
		label = "N/A";
		
		//BigInteger millis = BigInteger.valueOf(Instant.now().toEpochMilli());
		//updated_at = millis.multiply(BigInteger.valueOf(1000000L));
		updated_at = BigInteger.valueOf(0L);
	}
	
	public StateLocation(String label) {
		location = new byte[16];
		new Random().nextBytes(location);
		
		this.label = label;
		
		//BigInteger millis = BigInteger.valueOf(Instant.now().toEpochMilli());
		//updated_at = millis.multiply(BigInteger.valueOf(1000000L));
		updated_at = BigInteger.valueOf(0L);
	}
	
	public StateLocation(SetLocation setLocation) {
		location = setLocation.location;
		
		label = setLocation.label;
		
		BigInteger millis = BigInteger.valueOf(Instant.now().toEpochMilli());
		updated_at = millis.multiply(BigInteger.valueOf(1000000L));
	}
	
	public int getCode() {
		return code;
	}
	
	public byte[] getLocation() {
		return location;
	}
	
	public void setLocation(byte[] location) {
		this.location = location;
	}
	
	public String getLabel() {
		return label;
	}
	
	public void setLabel(String label) {
		this.label = label;
	}
	
	public BigInteger getUpdatedAt() {
		return updated_at;
	}
	
	public void setUpdatedAt(BigInteger updated_at) {
		this.updated_at = updated_at;
	}
	
	public void setUpdatedAtToNow() {
		BigInteger millis = BigInteger.valueOf(Instant.now().toEpochMilli());
		updated_at = millis.multiply(BigInteger.valueOf(1000000L));
	}
	
	public void setFromCommandByteArray(byte[] byteArray) {
		for(int i=51; i>35; i--) location[(-1*i)+51] = byteArray[i];
		
		byte[] labelBytes = new byte[32];
		for(int i=52; i<84; i++) labelBytes[i-52] = byteArray[i];
		label = new String(labelBytes);
		
		//String updatedAtBinStr = "";
		//for(int i=91; i>83; i--) updatedAtBinStr = updatedAtBinStr.concat(CommonMethods.convertByteToBinaryString(byteArray[i]));
		//updated_at = BigInteger.valueOf(Long.parseLong(updatedAtBinStr, 2));
		byte[] updatedAtBytes = new byte[8];
		for(int i=91; i>83; i--) updatedAtBytes[(-1*i)+91] = byteArray[i];
		updated_at = new BigInteger(updatedAtBytes);
	}

}

File: src/main/java/org/stevenlowes/tools/lifxcontroller/commands/response/device/StateLocation.kt
package org.stevenlowes.tools.lifxcontroller.commands.response.device

import org.stevenlowes.tools.lifxcontroller.Utils
import org.stevenlowes.tools.lifxcontroller.commands.response.ResponseCommandUpdateTime

import java.math.BigInteger

data class StateLocation(val location: ByteArray = Utils.randomBytes(16),
                    val label: String = "N/A") : ResponseCommandUpdateTime(50) {

    companion object {
        fun loadFrom(byteArray: ByteArray): StateLocation{
            val location = ByteArray(16)
            for (i in 51 downTo 36) {
                location[-1 * i + 51] = byteArray[i]
            }

            val labelBytes = ByteArray(32)
            System.arraycopy(byteArray, 52, labelBytes, 0, 32)
            val label = String(labelBytes)

            //String updatedAtBinStr = "";
            //for(int i=91; i>83; i--) updatedAtBinStr = updatedAtBinStr.concat(Utils.convertByteToBinaryString(payloadBytes[i]));
            //updated_at = BigInteger.valueOf(Long.parseLong(updatedAtBinStr, 2));
            val updatedAtBytes = ByteArray(8)
            for (i in 91 downTo 84) {
                updatedAtBytes[-1 * i + 91] = byteArray[i]
            }
            val updatedAt = BigInteger(updatedAtBytes)

            val stateLocation = StateLocation(location, label)
            stateLocation.updatedAt = updatedAt
            return stateLocation
        }
    }
}
